src/consumer/stream_consumer.rs

```diff
@@ -167,29 +154,24 @@ where
 {
     type Item = KafkaResult<BorrowedMessage<'a>>;
 
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        // Unconditionally store the waker so that we are woken up if the queue
-        // flips from non-empty to empty. We have to store the waker on every
-        // call to poll in case this future migrates between tasks. We also need
-        // to store the waker *before* calling poll to avoid a race where `poll`
-        // returns None to indicate that the queue is empty, but the queue
-        // becomes non-empty before we've installed the waker.
-        self.set_waker(cx.waker().clone());
-
-        match self.poll() {
-            None => Poll::Pending,
-            Some(message) => Poll::Ready(Some(message)),
-        }
-    }
-}
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(notified) = &mut self.notified {
+                ready!(notified.as_mut().poll(cx));
+                self.notified = None;
+            }
 
-impl<'a, C, R> Drop for MessageStream<'a, C, R>
-where
-    C: ConsumerContext + 'static,
-{
-    fn drop(&mut self) {
-        let mut wakers = self.context().wakers.lock().expect("lock poisoned");
-        wakers.remove(self.slot);
+            match self.poll() {
+                None => {
+                    self.notified = Some(self.consumer.context().wakers.notified().boxed());
+                }
+                Some(message) => {
+                    // More messages were available, notify all other waiters
+                    self.consumer.context().wakers.notify_waiters();
+                    return Poll::Ready(Some(message));
```

Comment:
I don't think I get this. In the typical case, when the message queue becomes non-empty, we'll wake up precisely one waiter. Then we'll end up here, because we know there is at least one message available, at which point we'll notify all the wakers. So why not just call `wakers.notify_wakers` from the context to start?

Reply:
The problem with calling `notify_waiters` is that it will *only* notify the currently registered waiters, whereas `notify_one` will register one waiter or the next waiter, if none are currently waiting.

We also can't call both notify functions when the non-empty callback is called since it would be possible to race it such that

```
Task 1 sees no message, returns from poll
Non empty callback notifies all waiters
Task 1 acquires the handle from `notified()` but won't be woken up until the next non-empty callback.
```

Always notifying when we know there are likely messages solves this issue.

Reply:
Ah, ok, thanks for explaining.